portal/portal-service-impl/impl/src/java/org/sakaiproject/portal/service/BullhornServiceImpl.java

```diff
@@ -0,0 +1,388 @@
+                                            String url = assignmentService.getDeepLink(siteId, assignmentId);
+                                            doAcademicInsert(from, to, event, ref, title, siteId, e.getEventTime(), url);
+                                            countCache.remove(to);
+                                        }
+                                    }
+                                }
+                            } catch (IdUnusedException idue) {
+                                log.error("Failed to find either the assignment or the site", idue);
+                            }
+                        } else if (AssignmentConstants.EVENT_GRADE_ASSIGNMENT_SUBMISSION.equals(event)) {
+                            String siteId = pathParts[3];
+                            String submissionId = pathParts[pathParts.length - 1];
+                            try {
+                                SecurityAdvisor sa = unlock();
+                                AssignmentSubmission submission = assignmentService.getSubmission(submissionId);
+                                lock(sa);
+                                if (submission.getGradeReleased()) {
+                                    Site site = siteService.getSite(siteId);
+                                    Assignment assignment = submission.getAssignment();
+                                    String title = assignment.getTitle();
+                                    for (String to : ((List<String>) submission.getSubmitterIds())) {
+                                        switchUser(to);
+                                        String url = assignmentService.getDeepLink(siteId, assignment.getId());
+                                        doAcademicInsert(from, to, event, ref, title, siteId, e.getEventTime(), url);
+                                        countCache.remove(to);
+                                    }
+                                }
+                            } catch (IdUnusedException idue) {
+                                log.error("Failed to find either the submission or the site", idue);
+                            }
+                        }
+                    }
+                } catch (Exception ex) {
+                    log.error("Caught exception whilst handling events", ex);
+                }
+            }).start();
+        }
+    }
+
+    private SecurityAdvisor unlock() {
+
+        SecurityAdvisor securityAdvisor = new SecurityAdvisor() {
+                public SecurityAdvice isAllowed(String userId, String function, String reference) {
+                    return SecurityAdvice.ALLOWED;
+                }
+            };
+        securityService.pushAdvisor(securityAdvisor);
+        return securityAdvisor;
+    }
+
+    private void lock(SecurityAdvisor securityAdvisor) {
+        securityService.popAdvisor(securityAdvisor);
+    }
+
+    private String switchUser(String to) {
+
```

Comment:
This is going to be a bit tough. Profile2 always uses the current user when constructing its urls in the apis, hence the session switching. It's probably going to mean a refactor of Profile2's link generator which would then have to be backported to 11.x if people want to use the bullhorns there. Not a massive deal, but more work.

Thanks for all the reviews by the way, much appreciated.